edn/src/edn.rustpeg

```diff
@@ -252,6 +282,210 @@ pub entity -> Entity<ValueAndSpan>
+//
+// We expect every rule except the `raw_*` rules to eat whitespace
+// (with `__`) at its start and finish.  That means that every string
+// pattern (say "[") should be bracketed on either side with either a
+// whitespace-eating rule or an explicit whitespace eating `__`.
+
+query_function -> query::QueryFunction
+    = __ n:$(symbol_name) __ {? query::QueryFunction::from_symbol(&PlainSymbol::plain(n)).ok_or("expected query function") }
+
+fn_arg -> query::FnArg
+    = v:value {? query::FnArg::from_value(&v).ok_or("expected query function argument") }
+    / __ "[" args:fn_arg+ "]" __ { query::FnArg::Vector(args) }
+
+find_elem -> query::Element
+    = __ v:variable __ { query::Element::Variable(v) }
+    / __ "(" __ "the" v:variable ")" __ { query::Element::Corresponding(v) }
+    / __ "(" __ "pull" var:variable "[" patterns:pull_attribute+ "]" __ ")" __ { query::Element::Pull(query::Pull { var, patterns }) }
+    / __ "(" func:query_function args:fn_arg* ")" __ { query::Element::Aggregate(query::Aggregate { func, args }) }
+
+find_spec -> query::FindSpec
+    = f:find_elem "." __ { query::FindSpec::FindScalar(f) }
+    / fs:find_elem+ { query::FindSpec::FindRel(fs) }
+    / __ "[" f:find_elem __ "..." __ "]" __ { query::FindSpec::FindColl(f) }
+    / __ "[" fs:find_elem+ "]" __ { query::FindSpec::FindTuple(fs) }
+
+pull_attribute -> query::PullAttributeSpec
+    = __ "*" __ { query::PullAttributeSpec::Wildcard }
+    / __ k:raw_forward_namespaced_keyword __ alias:(":as" __ alias:raw_forward_keyword __ { alias })? {
+        let attribute = query::PullConcreteAttribute::Ident(::std::rc::Rc::new(k));
+        let alias = alias.map(|alias| ::std::rc::Rc::new(alias));
+        query::PullAttributeSpec::Attribute(
+            query::NamedPullAttribute {
+                attribute,
+                alias: alias,
+            })
+    }
+
+limit -> query::Limit
+    = __ v:variable __ { query::Limit::Variable(v) }
+    / __ n:(raw_octalinteger / raw_hexinteger / raw_basedinteger / raw_integer) __ {?
+        if n > 0 {
+            Ok(query::Limit::Fixed(n as u64))
+        } else {
+            Err("expected positive integer")
+        }
+    }
+
+order -> query::Order
+    = __ "(" __ "asc" v:variable ")" __ { query::Order(query::Direction::Ascending, v) }
+    / __ "(" __ "desc" v:variable ")" __ { query::Order(query::Direction::Descending, v) }
+    / v:variable { query::Order(query::Direction::Ascending, v) }
+
+
+pattern_value_place -> query::PatternValuePlace
+    = v:value {? query::PatternValuePlace::from_value(&v).ok_or("expected pattern_value_place") }
+
+pattern_non_value_place -> query::PatternNonValuePlace
+    = v:value {? query::PatternNonValuePlace::from_value(&v).ok_or("expected pattern_non_value_place") }
+
+pattern -> query::WhereClause
+    = __ "["
+      src:src_var?
+      e:pattern_non_value_place
+      a:pattern_non_value_place
+      v:pattern_value_place?
+      tx:pattern_non_value_place?
```

Comment:
I haven't looked closely, but I wonder if this sequence of optionals will always work correctly, and never leave `v` unbound while binding `tx`…

Reply:
I do not think it is possible, because everything that can come in a non-value place can also come in a value-place.  I'm not sure it's worth trying to be more careful here.